### PR TITLE
Rewrite TF batch norm; remove batch_norm_inference

### PR DESCRIPTION
**Pull request by CharlieL7**
* Depends on #1362
* Rewrites the TF batch norm like operators to other MIGX operators
* Removes the code related to batch_norm_inference
* Do we have an equivalent to `verify_onnx.cpp` for the TF parser?

### CODE REVIEW

**Comment by TedThemistokleous** [comment]
@CharlieL7  can you target this on the branch it depents on from PR #1362 , it'll be easier to see changes relative to that branch

File: src/tf/parse_batchnorm.cpp [hunk 1]
```diff
@@ -38,16 +39,44 @@ struct parse_batchnorm : op_parser<parse_batchnorm>
     instruction_ref parse(const op_desc& /*opd*/,
                           const tf_parser& /*parser*/,
                           tf_parser::node_info info,
-                          const std::vector<instruction_ref>& args) const
+                          std::vector<instruction_ref> args) const
     {
-        float epsilon  = 1e-5f;
-        float momentum = 0.9f;
+        // different default epsilon than from ONNX
+        float epsilon = 1e-4f;
```
Comment (TedThemistokleous) [inline src/tf/parse_batchnorm.cpp:45]: Why is the tolerance lower on MIGraphX vs Onnx? Was that a specific design call made or were we hitting some error without this?

File: test/tf/gen_tf_pb.py [hunk 2]
```diff
@@ -142,19 +144,21 @@ def batchnormv3_test(g1):
...
                                                shape=(32),
-                                               name='4')
+                                               name='variance')
         tf.raw_ops.FusedBatchNormV3(x=g1_input,
                                     scale=g1_scale,
                                     offset=g1_offset,
                                     mean=g1_mean,
                                     variance=g1_variance,
-                                    epsilon=0.00001,
+                                    epsilon=1e-6,
```
Comment (TedThemistokleous) [inline test/tf/gen_tf_pb.py:161]: We're more precice for this test but less on the other test for batchnorm_test() above, why is that?

**Review by TedThemistokleous — COMMENTED** [review]
Few comments on this one. Just curious mainly about the tolerance. Looks like you removed the batch_norm_inference and unrolled everything accordingly as part of the tests. which is nice.

File: src/tf/parse_batchnorm.cpp (hunk 1 above)
Comment (CharlieL7) [reply src/tf/parse_batchnorm.cpp:45]: This is just a different between the TensorFlow spec and the ONNX spec.

File: test/tf/gen_tf_pb.py (hunk 2 above)
Comment (CharlieL7) [reply test/tf/gen_tf_pb.py:161]: Testing that the epsilon value is being passed over correctly.

File: src/tf/parse_batchnorm.cpp (hunk 1 above)
Comment (TedThemistokleous) [reply src/tf/parse_batchnorm.cpp:45]: Oh okay, so there's no way around it. Got it

File: test/tf/gen_tf_pb.py (hunk 2 above)
Comment (TedThemistokleous) [reply test/tf/gen_tf_pb.py:161]: Oh that makes sense.

**Comment by TedThemistokleous** [comment]
Not sure why CI is being odd on the perf test again. Everything else looks fine for this one.

**Review by TedThemistokleous — APPROVED** [review]

File: src/tf/parse_batchnorm.cpp (hunk 1 above)
Comment (umangyadav) [reply src/tf/parse_batchnorm.cpp:45]: Are you using, this https://www.tensorflow.org/api_docs/python/tf/keras/layers/BatchNormalization 

or this : https://www.tensorflow.org/api_docs/python/tf/nn/batch_normalization

I assume, keras one. It uses epsilon = 0.001 == 1e-3f.

File: test/tf/tf_test.cpp [hunk 3]
```diff
@@ -186,50 +185,87 @@ TEST_CASE(batchmatmul_test)
...
+    auto denom    = mm->add_instruction(migraphx::make_op("pow"), var_eps, mb_rt);
+    auto mb_denom = mm->add_instruction(
+        migraphx::make_op("multibroadcast", {{"out_lens", {1, 32, 16, 16}}}), denom);
+    auto div0     = mm->add_instruction(migraphx::make_op("div"), numer, mb_denom);
+    auto mb_scale = mm->add_instruction(
+        migraphx::make_op("multibroadcast", {{"out_lens", {1, 32, 16, 16}}}), usq_scale);
+    auto r0      = mm->add_instruction(migraphx::make_op("mul"), div0, mb_scale);
+    auto mb_bias = mm->add_instruction(
+        migraphx::make_op("multibroadcast", {{"out_lens", {1, 32, 16, 16}}}), usq_bias);
+    mm->add_instruction(migraphx::make_op("add"), r0, mb_bias);
```
Comment (umangyadav) [inline test/tf/tf_test.cpp:223]: This expansion of batchNorm appears in a few places. Would it be better if we add a helper method `add_bn` in `common.hpp` ?

File: test/tf/gen_tf_pb.py [hunk 4]
```diff
@@ -120,19 +120,21 @@ def batchnorm_test(g1):
     with g1.as_default():
         g1_input = tf.compat.v1.placeholder(tf.float32,
                                             shape=(1, 16, 16, 32),
-                                            name='0')
-        g1_scale = tf.constant(1.0, dtype=tf.float32, shape=[32], name='1')
-        g1_offset = tf.compat.v1.placeholder(tf.float32, shape=(32), name='2')
-        g1_mean = tf.compat.v1.placeholder(tf.float32, shape=(32), name='3')
+                                            name='x')
+        g1_scale = tf.constant(1.0, dtype=tf.float32, shape=[32], name='scale')
```
Comment (umangyadav) [inline test/tf/gen_tf_pb.py:124]: Would be good to have `fp16` test and a test that would hit `unsqueeze` lines for batch norm parameters.

**Review by umangyadav — APPROVED** [review]

File: src/driver/resnet50.cpp [hunk 5]
```diff
@@ -25,1453 +25,912 @@
...
-    auto x_main_module_12 = mmain->add_literal(migraphx::abs(
-        migraphx::generate_literal(migraphx::shape{migraphx::shape::float_type, {512}}, 11)));
-    auto x_main_module_13 = mmain->add_literal(migraphx::generate_literal(
-        migraphx::shape{migraphx::shape::float_type, {512, 512, 3, 3}}, 12));
-    auto x_main_module_14 = mmain->add_literal(migraphx::abs(
-        migraphx::generate_literal(migraphx::shape{migraphx::shape::float_type, {512}}, 13)));
-    auto x_main_module_15 = mmain->add_literal(
-        migraphx::generate_literal(migraphx::shape{migraphx::shape::float_type, {512}}, 14));
+        migraphx::generate_literal(migraphx::shape{migraphx::shape::float_type, {512}}, 4));
+    auto x_main_module_6 = mmain->add_literal(migraphx::generate_literal(
```
Comment (umangyadav) [inline src/driver/resnet50.cpp:49]: Not for this PR, but i guess we can write these models similar to, 
https://github.com/keras-team/keras-applications/blob/master/keras_applications/resnet50.py

In a modularized way.

File: test/tf/gen_tf_pb.py (hunk 4 above)
Comment (CharlieL7) [reply test/tf/gen_tf_pb.py:124]: Not sure I understand what you meant by hitting the `unsqueeze` lines.

File: test/tf/tf_test.cpp (hunk 3 above)
Comment (CharlieL7) [reply test/tf/tf_test.cpp]: Simplified the tests with `add_common_op`

File: test/tf/gen_tf_pb.py (hunk 4 above)
Comment (umangyadav) [reply test/tf/gen_tf_pb.py:124]: Never mind about `unsqueeze`.  But fp16, test is good to have.

File: test/verify/test_conv_bn.cpp [hunk 6]
```diff
@@ -37,19 +38,52 @@ struct test_conv_bn : verify_program<test_conv_bn>
...
+        auto c_len   = bn_lens.at(1);
+        auto mb_mean = mm->add_instruction(
+            migraphx::make_op("multibroadcast", {{"out_lens", bn_lens}}), usq_mean);
+        auto numer  = mm->add_instruction(migraphx::make_op("sub"), conv, mb_mean);
+        auto mb_eps = mm->add_instruction(
+            migraphx::make_op("multibroadcast", {{"out_lens", {c_len, 1, 1}}}), eps);
+        auto var_eps = mm->add_instruction(migraphx::make_op("add"), usq_var, mb_eps);
+        auto mb_rt   = mm->add_instruction(
+            migraphx::make_op("multibroadcast", {{"out_lens", {c_len, 1, 1}}}), rt);
+        auto denom    = mm->add_instruction(migraphx::make_op("pow"), var_eps, mb_rt);
```
Comment (umangyadav) [inline test/verify/test_conv_bn.cpp:77]: Use `add_common_op` similar to other test ?

File: src/driver/inceptionv3.cpp [hunk 7]
```diff
@@ -35,2604 +36,1595 @@ migraphx::program inceptionv3(unsigned batch) // NOLINT(readability-function-siz
     migraphx::module_ref mmain = p.get_main_module();
     auto x_main_module_0       = mmain->add_literal(migraphx::abs(
         migraphx::generate_literal(migraphx::shape{migraphx::shape::float_type, {1}}, 0)));
-    auto x_0                   = mmain->add_parameter(
-        "0", migraphx::shape{migraphx::shape::float_type, {batch, 3, 299, 299}});
+    auto x_input               = mmain->add_parameter(
```
Comment (umangyadav) [inline src/driver/inceptionv3.cpp:39]: Have you run each of these three driver models through `migraphx-driver verify` ?

Comment (CharlieL7) [reply src/driver/inceptionv3.cpp:39]: Looks like the models are good to go, no updates.

**Comment by CharlieL7** [comment]
For the epsilon value I'm using the eps from this: https://www.tensorflow.org/api_docs/python/tf/raw_ops/FusedBatchNorm and this: https://www.tensorflow.org/api_docs/python/tf/raw_ops/FusedBatchNormV3. Since those are the TensorFlow ops used in the tests.

**Review by umangyadav — APPROVED** [review]
If you can add FP16 test, that would be good. Otherwise LGTM
